lib/THC/THCTensorMode.cu

```diff
@@ -0,0 +1,202 @@
+#include "THC.h"
+#include "THCThrustAllocator.cuh"
+#include <thrust/device_ptr.h>
+#include <thrust/sort.h>
+#include <thrust/inner_product.h>
+#include <thrust/device_vector.h>
+#include <thrust/extrema.h>
+#include <thrust/execution_policy.h>
+#include <thrust/sequence.h>
+
+// Used to do []-style selection of an input Tensor given a storage of dimension values - i.e.
+//
+// multiSelect(state, self, src, {2, 0, 3})
+//
+// Should be equivalent to a Torch level operation: self = src[2][0][3]. In particular, we want
+// to leave self as a single vector, so the number of positions to index through must be one less
+// than the number of dimensions in src.
+THC_API void THCudaTensor_multiSelect(THCState* state, THCudaTensor* self, THCudaTensor* src, THLongStorage *position) {
+  THAssert(THLongStorage_size(position) == THCudaTensor_nDimension(state, src) - 1);
+
+  THCudaTensor* temp = src;
+  for (int i = 0; i < THLongStorage_size(position); ++i) {
+    THCudaTensor_select(state, self, temp, 0, THLongStorage_data(position)[i]);
+    temp = self;
+  }
+}
+
+THC_API void THCudaTensor_calculateMode(THCState* state,
+                                        THCudaTensor* values,
+                                        THCudaLongTensor* indices,
```

Comment:
@soumith I think this should just be a CPU Tensor for the indices, but I wasn't sure. What do you think?